azure-iot-device/azure/iot/device/patch.py

```diff
@@ -0,0 +1,133 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+"""This module provides patches used to dynamically modify items from the libraries"""
+
+import sys
+import inspect
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+# TODO: make this work for Python 2.7 and 3.4
+def add_shims_for_inherited_methods(target_class):
+    """Dynamically add overriding, pass-through shim methods for all public inherited methods
+    on a child class, which simply call into the parent class implementation of the same method.
+
+    These shim methods will include the same docstrings as the method from the parent class.
+
+    This currently only works for Python 3.5+
+
+    :param target_class: The child class to add shim methods to
+    """
+
+    # Depending on how the method was defined, it could be either a function or a method.
+    # Thus we need to find the union of the two sets.
+    # Here instance methods are considered functions because they are not yet bound to an instance
+    # of the class. Classmethods on the other hand, are already bound, and show up as methods.
+    # It also is worth noting that async functions/methods ARE picked up by this introspection.
+    class_functions = inspect.getmembers(target_class, predicate=inspect.isfunction)
+    class_methods = inspect.getmembers(target_class, predicate=inspect.ismethod)
+    all_methods = class_functions + class_methods
+
+    # This list of attributes gives us a lot of information, but we only are using it to get
+    # the defining class of a given method.
+    class_attributes = inspect.classify_class_attrs(target_class)
+
+    # We must alias classnames to prevent naming collisions when this fn is called multiple times
+    # with classes that share a name. If we've already used this classname, add trailing underscore(s)
+    classname_alias = target_class.__name__
+    while classname_alias in globals():
+        classname_alias += "_"
+
+    # Import the class we're adding methods to, so that functions defined in this scope can use super()
+    class_module = inspect.getmodule(target_class)
+    exec(
+        "from {module} import {target_class} as {alias}".format(
+            module=class_module.__name__, target_class=target_class.__name__, alias=classname_alias
+        ),
+        globals(),
```

Comment:
>       globals(), [](start = 1, length = 17)

why do you have to specify globals here?  The docs I've found are fairly vague.  Isn't this the default?

Reply:
When `exec` is called without any `globals` or `locals` optional args (side note: I need to change the syntax to represent this is an optional arg), exec operates within the current scope, i.e. the local scope of this function `add_shims_for_inherited_methods`. You can override the default scope by providing your own `globals` or `locals`. In this case, we don't bother with `locals` because changes made there aren't reflected - think of it as read-only. Since we're using `exec` to actually define new things, we need to use a globals scope.

For more information, read this doc: [https://docs.python.org/3/library/functions.html#exec](https://docs.python.org/3/library/functions.html#exec)

---
In reply to: [315357730](https://github.com/Azure/azure-iot-sdk-python-preview/pull/164#discussion_r315357730) [](ancestors = 315357730)